jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/AbstractNodeHandler.java

```diff
@@ -587,17 +589,26 @@ protected String checkSignalAndConvertToRealSignalNam(ExtensibleXmlParser parser
         }
         signalNames.add(signalName);
 
-        Map<String, Signal> signals = (Map<String, Signal>) buildData.getMetaData("Signals");
+        Map<String, Signal> signals = (Map<String, Signal>) buildData.getMetaData(SIGNALS);
         if (signals != null ) {
             if( signals.containsKey(signalName)) {
                 Signal signal = signals.get(signalName);
                 signalName = signal.getName();
                 if (signalName == null) {
                     throw new IllegalArgumentException("Signal definition must have a name attribute");
                 }
+            } else {
+                // align signals metadata with signal names 
+                // (for backward compatibility we need to support signals even if not defined)
+                signals.put(signalName, new Signal(signalName, signalName, null));
```

Comment:
Our engine is supporting undefined signals (meaning that you can put a signal name in an intermediate event without a signal definition). I do not want to enforce that (because process that are now considered validated, wont be anymore), so I have to add this comparison in code

Reply:
Any time you create a signal in the bpmn2 modeler should create a signal element. So this should not be needed.  
This is actually hiding a problem in the bpmn2 modeler IMO if that is happening.

There are few cases where signals are used to implement something else. They should not be defined as signals but messages.
Start Message (implemented as signal)
Boundary events (messages BoundaryEventHandler)
catching events (messages)

this should be removed.



In spite they are implemented as signal you need to make a difference in objects there.

Reply:
There are some processes with have a signal ref, but not define signal object, this code is trying to completate that scenario
https://github.com/kiegroup/jbpm/blob/master/jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/signal.bpmn

Reply:
After talking with Enrique, this non defined signals should not be exposed externally